### PR TITLE
feat: split demo and migrate to rollup

### LINKED ISSUE #179: Split the demo app to a different repository
The goal of this issue is to separate the demo app from this repository for the following reasons:

1. Migrate the SDK building framework to `rollup` for better flexibility.
2. Compile the demo with either `parcel` or `vite`, which allows more straightforward configuration for fully-fledged SPA apps.
3. Prevent polluting the changelog with demo changes, enabling the use of other tags in conventional commits, simplifying the lifecycle of both repositories.

-----

While working on issue #175, we observed that the build process has become overly complex. The following issues have been identified:

- **UMD File Generation with Parcel:** Generating a UMD-compliant file with Parcel seems challenging unless explicit window settings are made in a dedicated entry file, as illustrated below:

```javascript
import './src/middleware/srgssr.js';
import './src/components/player.js';
import Pillarbox from './src/pillarbox.js';

window.Pillarbox = Pillarbox;
```

We have tested the following alternatives:

## Vite

It is easy to generate both ES and UMD modules with Vite. The demo page is handled correctly through multiple configuration files. However, dependency exclusion in the final build with Vite requires specifying each excluded dependency or building the library in SSR (Server Side Rendering) mode. See [Building in library mode for node instead of browser](https://github.com/vitejs/vite/issues/13926#issuecomment-1686013974) This is a working example with SSR mode :

```javascript
export default defineConfig({
    build: {
        ssr: true,
        rollupOptions: {
            input: 'src/main.js',
            output: {
                entryFileNames: 'pillarbox-core.js',
            }
        }
    }
});
```

## Rollup

Configuring Rollup for multiple targets in the SDK part is straightforward, offering extensive customization options. However, configuring the demo, especially serving it in a development server, becomes convoluted with multiple plugins and complex configurations. See this WIP configuration for the demo part:

```javascript
const isServe = process.env.SERVE === 'true';

export default [
    {
        input: 'demo/*.html',
        output: {
            dir: 'dist',
            format: 'esm',
            entryFileNames: '[name].[hash].js',
            assetFileNames: '[name].[hash][extname]'
        },
        plugins: [
            commonjs(),
            resolve(),
            terser(),
            scss({
                output: true,
                minify: true
            }),
            html(),
            importMetaAssets(),
            isServe && serve('demo'),
            isServe && livereload()
        }
    }
];
```

Both solutions are not satisfying to contain the demo and SDK in the same repository, hence the decision to split them.

### PR DESCRIPTION
**Pull request by jboix**
## Description

Closes #179 by migrating the compilation framework from `Parcel` to `Rollup`, aiming to improve the flexibility of the compilation process. To execute the full bundle of the library simply run: `npm run build`.

#### `Rollup` migration

[Rollup](https://rollupjs.org/introduction/) is the bundler that allows us to create an UMD package of the library, which can be used in various environments. The proposed configuration produces two main builds:

- `pillarbox-core`: This build includes only the superset of functions of video.js, the underlying video player library. It does not include analytics or data providers that are specific to SRG/SSR.
- `pillarbox`: This build includes both the core and the specific business logic for SRG/SSR content, i.e.: The analytics and the data provider.

Both these builds are provided in UMD and ES format.

The following plugins have been introduced:

- `@rollup/plugin-commonjs`: converts CommonJS modules to ES6, so they can be included in a Rollup bundle.
- `@rollup/plugin-json`: converts JSON files to ES6 modules, so they can be imported as objects.
- `@rollup/plugin-node-resolve`: locates modules using the Node resolution algorithm, for using third party modules in node_modules.
- `@rollup/plugin-terser`: generates a minified bundle with terser, a JavaScript optimizer tool.
- `rollup-plugin-filesize`: shows the filesize of the bundle in the cli.

The following command executes the generation of both these bundles: `npm run build:lib`.

#### Introduction of `sass`

As part of the migration the preprocessing of `scss` files is now done directly with [Sass](https://sass-lang.com/install/) as opposed of `parcel`. A minified `css` stylesheet and a sourcemap are generated by executing: `npm run build:css`.

#### Add typings generation via typescript

TypeScript is a superset of JavaScript that adds static type checking and other features. It helps to write more robust and error-free code. Even though we've opted to not use typescript in `Pillarbox` a `tsconfig.json` file has been added to generate the declaration files based on the JSDocs of the source code.

Executing: `npm run build:typings` will generate the typings declarations.

See [Creating .d.ts Files from .js files](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html).

#### Split the demo application to its own repository

The demo folder has been moved to SRGSSR/pillarbox-web-demo where it can be maintained and updated independently. The following pull request descibres the changes that have taken place on the demo part: SRGSSR/pillarbox-web-demo#1

Since the demo has moved to another respository the github page of this one is now a standalone player.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -5,9 +5,7 @@
   <meta charset="UTF-8" />
   <meta name="viewport" content="width=device-width, initial-scale=1.0" />
   <title>Pillarbox - Development Page</title>
-  <link rel="icon" type="image/x-icon" href="img/favicon.ico">
```
Comment (amtins) [inline index.html:8]: The demo page doesn't deserve to have the Pillarbox favicon ? :(

File: package.json [hunk 2]
```diff
@@ -93,20 +58,18 @@
...
-    "parcel": "^2.10.3",
+    "parcel": "^2.11.0",
     "rimraf": "^5.0.5",
+    "rollup": "^4.9.6",
+    "rollup-plugin-filesize": "^10.0.0",
+    "sass": "^1.70.0",
     "semantic-release": "^22.0.10",
     "stylelint": "^16.0.2",
     "stylelint-config-rational-order": "^0.1.2",
-    "stylelint-order": "^6.0.4"
```
Comment (amtins) [inline package.json:109]: Secretly removed?

Comment (jboix) [reply package.json:109]: Good catch, I was testing some stuff for the demo page and forgot to put it back.
> Or maybe I secretly hate that it is always complaining about the order of my CSS attributes...

**Comment by amtins** [comment]
@jboix I think we need to find another strategy for this PR. This PR deals with two very distinct subjects, so it would be better to separate it into two. Once the two PRs are separated, I think it would be best to delete all commits related to the demo.

**Review by amtins — APPROVED** [review]
Looks fantastic! Thanks 🥇

**Pull request merged**